DOC/API: Styler documentation changes

TomAugspurger opened the pull request:
Closes https://github.com/pydata/pandas/issues/13222, #13225
- Update docstring and notebook on output shapes matching
- Change build process to use nbconvert to execute the notebook and then convert using NBConvert.

Second commit puts some more restrictions on the output of a user's function passed to `.apply`.

TomAugspurger committed:
DOC: Styler documentation changes

- Update docstring on output shapes matching
- Change build process

API: Enforce output shape requirements on Styler.apply

jreback commented on pandas/formats/style.py:
```diff
@@ -444,15 +444,17 @@ def apply(self, func, axis=0, subset=None, **kwargs):
-            to the entire DataFrame at once with ``axis=None``.
-        subset: IndexSlice
+            to the entire DataFrame at once with ``axis=None``
+        subset : IndexSlice
```
I thought we don't have a space before the :

? in general I mean

TomAugspurger replied:
The examples here do: https://github.com/numpy/numpy/blob/master/doc/HOWTO_DOCUMENT.rst.txt#sections, but I don't think it's enforced, does pandas have a preference?

jreback replied:
not sure we are consistent as have seen both ways (or does it even format differently)?

jreback replied:
I am wrong - it does specify spaces on both sides
(contrast this to linting which doesn't have s space - but different issue)

jorisvandenbossche replied:
yes, it definitly specify spaces on both sides, it's a comment I often give when reviewing docstrings ... :-) (and it is needed to get the html formatting correctly using numpydoc, don't know if napoleon is more flexible)

jreback replied:
maybe we should have a commit hook to check this? or lint check or something?

jreback replied:
https://gitlab.com/pycqa/flake8-docstrings anyone try?

jorisvandenbossche replied:
Last time I looked for something like this, it only checked for pep257, and not for numpydoc specific things

jorisvandenbossche replied:
But there seems some discussion/work under way: https://github.com/PyCQA/pydocstyle/issues/129

jreback replied:
yes that looks good / we should follow that release

jreback commented on pandas/formats/style.py:
```diff
@@ -461,9 +463,22 @@ def apply(self, func, axis=0, subset=None, **kwargs):
+        The output shape of ``func`` should match the input, i.e. if
+        ``x`` is the input row, column, or table (depending on ``axis``),
+        then ``func(x.shape) == x.shape`` should be true.
+
```
this is very easy to assert / check

you should just enforce it

TomAugspurger replied:
I'll push commit tonight with this (very easy, two line change). I had written a test that violated the rule, but "worked" 😳 . Just need to write additional tests.

jorisvandenbossche commented on doc/make.py:
```diff
@@ -102,10 +105,88 @@ def clean():
+    ep = ExecutePreprocessor(allow_errors=allow_errors,
+                             timeout=timeout,
+                             kernel_name=kernel_name)
+    ep.preprocess(nb, {'metadta': {'path': 'notebooks/'}})
```
metadta -> metadata (side question: what does this do?)

TomAugspurger replied:
```
resources : dictionary
    Additional resources used in the conversion process. For example,
    passing ``{'metadata': {'path': run_path}}`` sets the
    execution path to ``run_path``.
```

But apparently nbconvert doesn't do any argument checking :) I'll remove that.

jorisvandenbossche commented on pandas/formats/style.py:
```diff
@@ -444,15 +444,17 @@ def apply(self, func, axis=0, subset=None, **kwargs):
-        func: function
-        axis: int, str or None
+        func : function
+            ``func`` should take a Series or DataFrame (depending
```
And if we are doing nitpicky :-) The numpydoc standard says that references to keywords should have single backticks ...

TomAugspurger replied:
Ah, I thought we decided to use double ticks since it looks nicer in our HTML :)

jorisvandenbossche commented:
@TomAugspurger Did you change a lot in the notebook itself? (Or other phrased: is it worth reviewing? Because online the diff is not shown)

TomAugspurger commented:
@jorisvandenbossche the notebook changes were

Added:

>For `Styler.applymap` your function should take a scalar and return a single string with the CSS attribute-value pair.
For `Styler.apply` your function should take a Series or DataFrame (depending on the axis parameter), and return a Series or DataFrame with an identical shape where each value is a string with a CSS attribute-value pair.

and

> And crucially the input and output shapes of `func` must match. If `x` is the input then ``func(x).shape == x.shape``.

The  reason the diff is so big is that I deleted all the output from the notebook committed to the repo.

jreback commented:
@TomAugspurger what about have a shape check when using ``.apply``? e.g. to guarntee (or otherwise raise)?

TomAugspurger commented:
> @TomAugspurger what about have a shape check when using .apply

I'll push that change tonight, it's pretty easy, just need tests.

jreback commented:
gr8! @TomAugspurger better to have an exception than documentation. Further no issues with restricting ``.apply``. No need to make this too general (like ``DataFrame.apply``) that just leads to blind pathways.

TomAugspurger commented:
Sorry for the delay, pushed changes.

1. The result of `Styler.apply(func, axis=0/1)` must have the same shape. The output is from `DataFrame.apply`, so we know we have a Series with the correct index labels.
2. The result of `Styler.apply(func, axis=None)` must be a DataFrame with the same index columns / labels (and so same shape as well).

I put the second one in place to avoid confusing output like

![screen shot 2016-05-20 at 7 14 39 am](https://cloud.githubusercontent.com/assets/1312546/15427414/8a2b50ee-1e5a-11e6-83ed-b28ee517d991.png)

Only the top row gets colored, even though the output was 2x2.

jreback commented:
xref: https://travis-ci.org/pydata/pandas/jobs/132906479#L1631

TomAugspurger committed:
TST: Additional Styler init and repr tests

TomAugspurger commented:
Sorry, forgot this was still open.

Here are the relevant changes to the notebook

>  Notice that the output shape of highlight_max matches the input shape, an array with len(s) items.

and

> When using ``Styler.apply(func, axis=None)``, the function must return a DataFrame with the same index and column labels.

and

> And crucially the input and output shapes of `func` must match. If `x` is the input then ``func(x).shape == x.shape``.

jreback commented on doc/README.rst:
```diff
@@ -81,7 +81,7 @@ have ``sphinx`` and ``ipython`` installed. `numpydoc
 follow the Numpy Docstring Standard (see above), but you don't need to install
 this because a local copy of ``numpydoc`` is included in the pandas source
-code.
+code. To build the Jupyter notebooks included in the documentation, you'll need `nbconvert <https://nbconvert.readthedocs.io/en/latest/>`_ and `nbformat <http://nbformat.readthedocs.io/en/latest/>`_.
```
we might also have this same documenation in ``contributing.rst`` so update there as well (as well as some info on what deps you need to build the docs)

jorisvandenbossche replied:
Maybe we can just include this file in contributing.rst, to not duplicate things (I thought it was done like that before)

jreback commented on doc/make.py:
```diff
@@ -102,10 +105,90 @@ def clean():
+
+
+def execute_nb(src, dst, allow_errors=False, timeout=1000, kernel_name=''):
+    '''
```
use """

jreback commented on doc/make.py:
```diff
@@ -102,10 +105,90 @@ def clean():
+
+
+def convert_nb(src, dst, to='html', template_file='basic'):
+    '''
```
same

jreback commented on doc/source/whatsnew/v0.18.2.txt:
```diff
@@ -141,6 +141,7 @@ API changes
 - Non-convertible dates in an excel date column will be returned without conversion and the column will be ``object`` dtype, rather than raising an exception  (:issue:`10001`)
 - An ``UnsupportedFunctionCall`` error is now raised if NumPy ufuncs like ``np.mean`` are called on groupby or resample objects (:issue:`12811`)
 - Calls to ``.sample()`` will respect the random seed set via ``numpy.random.seed(n)`` (:issue:`13161`)
+- ``Styler.apply`` is now stricter about the outputs your function must return. For ``axis=0`` or ``axis=1``, the output shape must be identical. For ``axis=None``, the output must be a DataFrame with identical columns and index labels. (:issue:`13222`)
```
more strict

jreback commented on pandas/formats/style.py:
```diff
@@ -444,15 +462,19 @@ def apply(self, func, axis=0, subset=None, **kwargs):
         ----------
-        func: function
-        axis: int, str or None
+        func : function
```
you can change parameters to ``func:`` (no space)

jorisvandenbossche commented on pandas/formats/style.py:
```diff
@@ -427,11 +427,29 @@ def _compute(self):
+            result = func(data, **kwargs)
+            if not isinstance(result, pd.DataFrame):
+                raise TypeError("Function {!r} must return a DataFrame when "
+                                "passed to `Styler.apply` with axis=None")
```
Isn't there a `.format(..)` missing? (given the `{!r}`)

jorisvandenbossche commented on pandas/formats/style.py:
```diff
@@ -427,11 +427,29 @@ def _compute(self):
+            if not (result.index.equals(data.index) and
+                    result.columns.equals(data.columns)):
+                msg = ('Result of {!r} must have identical index and columns '
+                       'as the input')
```
same here

jorisvandenbossche commented:
@TomAugspurger Does it build locally for you? (the docs) As on travis there is an error "InputError: [Errno 2] No such file or directory: 'source/html-styling.html'" (not sure if that means that something went wrong ..)

TomAugspurger committed:
format error messages

TomAugspurger commented:
Addressed all the comments, thanks. @jorisvandenbossche it does build locally, running from `pandas/doc`. Is that failure from before this pull request? I'll check this output of this Travis run.

jorisvandenbossche commented:
Let's see how the travis docs will look like after it is merged.
